### PR TITLE
Emit `<!--Start PulumiCodeChooser -->` instead of `{{% examples %}}`

### PR DESCRIPTION
**Pull request by guineveresaenger**
**IMPORTANT** This cannot be merged until the bridge has updated pulumi/pulumi to a version that includes [these changes](https://github.com/pulumi/pulumi/commit/26bdac74383c599a7c976d301822883f38ac9b29).

After that, merging will be safe.

**Note**: https://github.com/pulumi/pulumi/pull/15475 needs to be merged first for these changes to be safe.

This PR aims to remove the nested "Example Usage" logic from the bridge, [as discussed here](https://github.com/pulumi/registry/issues/3498).

The issue this is aimed at fixing is https://github.com/pulumi/pulumi-aws/issues/2624 but there should be a myriad of other issues that will be addressed by this change.

In a nutshell, this PR removes any `{{% example(s) }}` injection from the `Description` fields of a resource and instead detects any and all code conversions from TF/HCL --> pulumi languages, and surrounds them with an HTML comment. 
This is currently 
```
const (
	startPulumiCodeChooser = "<!--Start PulumiCodeChooser -->"
	endPulumiCodeChooser   = "<!--End PulumiCodeChooser -->"
)
```

Marking up any converted code blocks using the new format will serve _only_ as a marker for registry docsgen to surround this converted code with a language chooser.

I refrained from going full Markdown parser because I believe it would be an even larger rewrite. Instead, we detect code fences in the docstring input, and append text and converted code section to the output, with the following caveats:

- If a code block is part of a Section (i.e. has a Markdown header), if we encounter errors in code conversion, the new functionality strips the entire section automatically. Non-headered code blocks with such an error will also continue to be stripped.
- Using indices rather than line splitting allows us to get rid of the awkward `<break>` placeholders for the Import section.
- The Import section no longer needs special handling in `convertExamplesInner`: it is valid code and will be retained as such
- With a little bit of extra parsing of the code fence syntax highlighter, we can capture a few more cases of perfectly valid json or yaml that we have elided previously as "failure to convert from TF".
- With the removal of examples, titles (which never actually made it into the schema) are now obsolete. We will refer to resource paths when logging conversion failures. We have done so all along when there was no `exampleTitle`, and this change affects only the wording of logging output.
- The `isFrontMatter` variable was removed as well. To the best of my knowledge, Hugo front matter does not get piped through the bridge, and is generated by the registry instead.
- Handwritten docs continue to bypass `convertExamplesInner`, with an added caveat that both third party and our own docs may add the old example shortcodes. This change will allow us to [remove some of these](https://github.com/pulumi/pulumi-aws/blob/master/docs/resource/aws_acm_certificate_validation.examples.md). Registry docsgen will continue to support the shortcode way of doing things, since native providers conform to that expectation separately.

### CODE REVIEW

File: pkg/tfgen/docs.go [hunk 1]
```diff
@@ -1357,129 +1361,139 @@ func (g *Generator) convertExamplesInner(
...
+				currentBlock.start = i
+			}
 		}
+	}
 
-		isImportSection := false
-		header, wroteHeader := section[0], false
-		isFrontMatter, isExampleUsage := !strings.HasPrefix(header, "## "), header == "## Example Usage"
+	// Now we know where the headers and code fences are.
+	textStart := 0
```
Comment (VenelinMartinov) [inline pkg/tfgen/docs.go:1408]: There's quite a few interactions and context one needs to keep when reading here, would this also make sense to be split into a few functions?

File: pkg/tfgen/test_data/convertExamples/aws_lambda_function_out.md [hunk 2]
```diff
@@ -0,0 +1,1388 @@
...
+
+-> To give an external source (like an EventBridge Rule, SNS, or S3) permission to access the Lambda function, use the `aws_lambda_permission` resource. See [Lambda Permission Model][4] for more details. On the other hand, the `role` argument of this resource is the function's execution role for identity and access to AWS services and resources.
+
+## Example Usage
+
+### Basic Example
+
+<!--Begin TFConversion -->
```
Comment (VenelinMartinov) [inline pkg/tfgen/test_data/convertExamples/aws_lambda_function_out.md:17]: Does this require a PR in the registry or somewhere else before merging?

**Comment by VenelinMartinov** [comment]
Okay, I think you've opened https://github.com/pulumi/pulumi/pull/15475 to parse the new HTML comments, great!

File: pkg/tfgen/docs.go [hunk 3]
```diff
@@ -1325,26 +1329,26 @@ func (g *Generator) convertExamples(docs string, path examplePath, stripSubsecti
...
 
 	// Use coverage tracker: on by default.
 	cov := true
-	return g.convertExamplesInner(docs, path, stripSubsectionsWithErrors, g.convertHCL, cov)
+	return g.convertExamplesInner(docs, path, g.convertHCL, cov)
 }
 
 // The inner implementation of examples conversion is parameterized by convertHCL so that it can be
 // executed either normally or in symbolic mode.
+
```
Comment (iwahbe) [inline pkg/tfgen/docs.go:1342]: Erroneous diff?
```suggestion
```

File: pkg/tfgen/docs.go [hunk 4]
```diff
@@ -1325,26 +1329,26 @@ func (g *Generator) convertExamples(docs string, path examplePath, stripSubsecti
...
 	docs string,
 	path examplePath,
-	stripSubsectionsWithErrors bool,
 	convertHCL func(
-		e *Example, hcl, path, exampleTitle string, languages []string,
+		e *Example, hcl, path string, languages []string,
 	) (string, error),
 	useCoverageTracker bool,
 ) (result string) {
+
```
Comment (iwahbe) [inline pkg/tfgen/docs.go:1351]: Erroneous diff?
```suggestion
```

File: pkg/tfgen/docs.go [hunk 5]
```diff
@@ -1357,129 +1361,139 @@ func (g *Generator) convertExamplesInner(
...
 
-	for _, section := range splitGroupLines(docs, "## ") {
-		if len(section) == 0 {
-			continue
+	// Find code fences in the presented doc
+	type tfBlockIndex struct {
+		start       int
+		end         int
+		headerStart int // a value of -1 indicates there's no header
+	}
```
Comment (iwahbe) [inline pkg/tfgen/docs.go:1369]: A doc comment explaining exactly what these numbers mean would be great.

I'm assuming that `start` refers to the first "\`" in the "\`\`\`hcl" that opens the block, `end` refers to the last "\`" that closes the block (such that `text[end+1] == '\n'` in the common case, and `headerStart` refers to the opening "#" in the closest enclosing `## Some header` for the block.

File: pkg/tfgen/docs.go [hunk 6]
```diff
@@ -1357,129 +1361,139 @@ func (g *Generator) convertExamplesInner(
...
+
+	var inCodeBlock bool
+	codeFence := "```"
+	h2 := "##"
+	h3 := "###"
```
Comment (iwahbe) [inline pkg/tfgen/docs.go:1374]: Are we not going to finds headers that are not h2 and h3s?

File: pkg/tfgen/test_data/address_map/expected.json [hunk 7]
```diff
@@ -25,5 +25,5 @@
     "ips": "The set of IPs on the Address Map.",
     "memberships": "Zones and Accounts which will be assigned IPs on this Address Map."
   },
-  "Import": "## Import\n\n```sh\u003cbreak\u003e $ pulumi import MISSING_TOK example \u003caccount_id\u003e/\u003caddress_map_id\u003e \u003cbreak\u003e```\u003cbreak\u003e\u003cbreak\u003e"
+  "Import": "## Import\n\n```sh\n$ pulumi import MISSING_TOK example \u003caccount_id\u003e/\u003caddress_map_id\u003e\n```\n\n"
```
Comment (t0yv0) [inline pkg/tfgen/test_data/address_map/expected.json:28]: Remaining issue with MISSING_TOK? Should that be an error instead?

File: pkg/tfgen/docs.go [hunk 8]
```diff
@@ -1357,129 +1361,139 @@ func (g *Generator) convertExamplesInner(
...
+	var codeIndices []tfBlockIndex
+	var currentBlock tfBlockIndex
+	currentBlock.headerStart = -1
+
+	for i := 0; i < (len(docs) - len(codeFence)); i++ {
+
+		if inCodeBlock == true {
```
Comment (iwahbe) [inline pkg/tfgen/docs.go:1381]: I'd much rather that we represented `inCodeBlock` as a function call that parses to the end of a code block, rather then a variable here.

Then the logic would look something like this:

```go
for i := 0; i < (len(docs) - len(codeFence)); i++ {
    if block, ok := atCodeBlock(docs, i); ok {
        codeIndices = append(codeIndices, block)
        i = block.end + 1
        continue
    }
    if end, ok := atHeader(docs, i); ok {
        headerStart = i
        i = end+1
    }
}
```

File: pkg/tfgen/docs.go [hunk 9]
```diff
@@ -1357,129 +1361,139 @@ func (g *Generator) convertExamplesInner(
...
-		sectionStart, sectionEnd := "", ""
-		if isExampleUsage {
-			sectionStart, sectionEnd = "{{% examples %}}\n", "{{% /examples %}}"
+		// append non-code text to output
+		if !stripSection {
+			if hasHeader {
+				fprintf(output, docs[textStart:tfBlock.headerStart])
+			} else {
+				fprintf(output, docs[textStart:tfBlock.start])
+			}
```
Comment (iwahbe) [inline pkg/tfgen/docs.go:1422]:
```suggestion
			end := tfBlock.start
			// if the section has a header we append the header after trying to convert the code.
			if tfBlock.headerStart >= 0 && textStart < tfBlock.headerStart {
				end = tfBlock.headerStart
			}
			fprintf(output, docs[textStart:end])
```

File: pkg/tfgen/docs.go [hunk 10]
```diff
@@ -1357,129 +1361,139 @@ func (g *Generator) convertExamplesInner(
...
 		}
+		// find the actual start index of the code
+		nextNewLine := strings.Index(docs[tfBlock.start:tfBlock.end], "\n")
+		if nextNewLine == -1 {
+			// write the line as-is; this is an in-line fence
+			fprintf(output, docs[tfBlock.start:tfBlock.end]+"```")
+			writeTrailingNewline(output)
+		} else {
+			syntaxHighlight := docs[tfBlock.start : tfBlock.start+nextNewLine+1]
```
Comment (iwahbe) [inline pkg/tfgen/docs.go:1441]: Can we call this something more descriptive?
```suggestion
			codeFence := docs[tfBlock.start : tfBlock.start+nextNewLine+1]
```

File: pkg/tfgen/docs.go [hunk 11]
```diff
@@ -1357,129 +1361,139 @@ func (g *Generator) convertExamplesInner(
...
+						stripSection = true
+						stripSectionHeader = tfBlock.headerStart
 					} else {
-						fprintf(subsectionOutput, "\n%s", line)
+						// append any headers and following text first
+						if hasHeader {
+							fprintf(output, docs[tfBlock.headerStart:tfBlock.start])
+							writeTrailingNewline(output)
+						}
+						fprintf(output, "<!--Begin TFConversion -->")
```
Comment (iwahbe) [inline pkg/tfgen/docs.go:1472]: `"<!--Begin TFConversion -->"` should be a constant.

File: pkg/tfgen/docs.go (hunk 6 above)
Comment (guineveresaenger) [reply pkg/tfgen/docs.go:1374]: It is possible but not very likely. The previous version of this code certainly only dealt in h2 and h3. We can make this more generic if we find that we are not "catching" some code blocks.

Comment (guineveresaenger) [reply pkg/tfgen/docs.go:1374]: hah, I found a better answer for you 💀 https://github.com/pulumi/pulumi-terraform-bridge/blob/master/pkg/tfgen/docs.go#L590

File: pkg/tfgen/test_data/convertExamples/wavefront_dashboard_json_out.md [hunk 12]
```diff
@@ -806,18 +805,16 @@ resources:
...
-{{% /example %}}
-{{% /examples %}}
 
 ## Import
 
 Dashboard JSON can be imported by using the `id`, e.g.:
 
 ```sh
- $ pulumi import wavefront:index/dashboardJson:DashboardJson dashboard_json tftestimport
+$ pulumi import wavefront:index/dashboardJson:DashboardJson dashboard_json tftestimport
```
[the remark this answers is not in the document]
Comment (guineveresaenger) [reply pkg/tfgen/test_data/convertExamples/wavefront_dashboard_json_out.md:819]: it is! the site render looks the same with and without.

File: pkg/tfgen/test_data/address_map/expected.json (hunk 7 above)
Comment (guineveresaenger) [reply pkg/tfgen/test_data/address_map/expected.json:28]: I'm happy to file an issue here but it's out of scope for this change I think!

File: pkg/tfgen/docs.go [hunk 13]
```diff
@@ -1357,129 +1361,139 @@ func (g *Generator) convertExamplesInner(
...
+			if docs[i:i+len(codeFence)] == codeFence {
+				currentBlock.end = i
+				codeIndices = append(codeIndices, currentBlock)
+				// reset
+				currentBlock.start = 0
+				currentBlock.end = 0
+				inCodeBlock = false
+			}
+		} else {
+			// Keep track of header locations. These should never be inside code blocks.
```
[the remark this answers is not in the document]
Comment (guineveresaenger) [reply pkg/tfgen/docs.go]: What I was trying to say was "we never look for a header inside a code block" - describing behavior, not assumption.
The new code is a bit more clear - we never look for a header if we are inside a code block.

**Review by VenelinMartinov — APPROVED** [review]
lgtm

File: pkg/tfgen/docs.go [hunk 14]
```diff
@@ -1308,7 +1305,14 @@ func (g *Generator) convertExamples(docs string, path examplePath, stripSubsecti
 		strings.Contains(docs, "```go") || strings.Contains(docs, "```yaml") ||
 		strings.Contains(docs, "```csharp") || strings.Contains(docs, "```java") {
 		// we have explicitly rewritten these examples and need to just return them directly rather than trying
-		// to reconvert them. But we need to surround them in the examples shortcode for rendering on the registry
+		// to reconvert them.
+		//
+		//TODO: This only works if the incoming docs already have an {{% example }} shortcode, and if they are
```
[the remark this answers is not in the document]
Comment (iwahbe) [reply pkg/tfgen/docs.go:1315]: Can you open an issue for this and link it with the TODO.

**Review by iwahbe — APPROVED** [review]
LGTM

**Comment by t0yv0** [comment]
> Note: https://github.com/pulumi/pulumi/pull/15475 needs to be merged first for these changes to be safe.

I don't know about 1699.

**Comment by iwahbe** [comment]
> > Note: [pulumi/pulumi#15475](https://github.com/pulumi/pulumi/pull/15475) needs to be merged first for these changes to be safe.
> 
> I don't know about 1699.

I don't believe that it is blocked on #1699.

**Comment by VenelinMartinov** [comment]
Yup, I confirmed it with docs - https://github.com/pulumi/pulumi-terraform-bridge/issues/1699 is not a blocker but the bridge has yet to pick up https://github.com/pulumi/pulumi/pull/15475

**Review by VenelinMartinov — APPROVED** [review]
lgtm

**Pull request merged**